builtins: categorize the volatilities for each overload

otan opened the pull request:
Note I have reverted the original volatility PR in this commit. This is because
I didn't want to affect the "Impure" checks at the moment, and let youse
optimizer folks handle it later. All the generator/window/aggregate builtins
are "Impure", but "Immutable" according to Postgres.

The easiest way to vet all the builtins is to run this query:
```
select proname, proargtypes,  provolatile from pg_catalog.pg_proc;
```

Resolves #26582.

Phew, what a flex friday.

----

This commit adds volatilities to each overload we implement. For now,
these should match the ones in Postgres whenever possible. Note this is
currently not used anywhere.

We have also added two tests:
* one test that ensures each builtin has a volatility.
* one test that checks the volatilies against a dump of psql
volatilities. We can add a blacklist to this later if we decide to
differ, but for now it is safer to match.

Release note: None

otan committed:
Revert "sql: mark functions as Immutable/Stable/Volatile"

This reverts commit 5ffa594a18a700877539b6c5070a4c3645333e8e.

Reverting this commit to define volatility on Overloads. We may need the
`Impure` functionality to replace `proleakproof`, and it seems that are
defined slightly differently.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48617)
<!-- Reviewable:end -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>


Thank you for your sacrifice.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48617#-:-M6qzvGj-5Aptmy8oi9H:bkqnzur)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/sem/builtins/all_builtins_test.go, line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48617#-M6qywsa1JFClKvMB2Jn:-M6qywsa1JFClKvMB2Jo:b-y4w22j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4a805facc66121917ccbdda8ddde1089e427375/pkg/sql/sem/builtins/all_builtins_test.go#L47)):*
> ```Go
> // overloads for Volatility.
> // Dump command below:
> // COPY (SELECT proname, args, rettype, provolatile, proleakproof FROM (SELECT lhs.oid, proname, pg2.typname as rettype, ARRAY_AGG(pg1.typname) as args, provolatile, proleakproof FROM (select oid, proname, unnest(proargtypes) as typ, proargnames, prorettype, provolatile, proleakproof from pg_proc) AS lhs JOIN pg_type AS pg1 ON (lhs.typ = pg1.oid) JOIN pg_type AS pg2 ON (lhs.prorettype = pg2.oid) GROUP BY lhs.oid, proname, pg2.typname, provolatile, proleakproof) a ORDER BY proname, args) TO '/tmp/pg_proc_provolatile_dump.csv' WITH CSV DELIMIER '|' HEADER;
> ```

[nit] break this up into multiple lines
___
*[pkg/sql/sem/builtins/all_builtins_test.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48617#-M6qyuLo1GQ262eJg-xb:-M6qyuLpDBE3YpfDF1se:bbeoxay) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4a805facc66121917ccbdda8ddde1089e427375/pkg/sql/sem/builtins/all_builtins_test.go#L48)):*
> ```Go
> // Dump command below:
> // COPY (SELECT proname, args, rettype, provolatile, proleakproof FROM (SELECT lhs.oid, proname, pg2.typname as rettype, ARRAY_AGG(pg1.typname) as args, provolatile, proleakproof FROM (select oid, proname, unnest(proargtypes) as typ, proargnames, prorettype, provolatile, proleakproof from pg_proc) AS lhs JOIN pg_type AS pg1 ON (lhs.typ = pg1.oid) JOIN pg_type AS pg2 ON (lhs.prorettype = pg2.oid) GROUP BY lhs.oid, proname, pg2.typname, provolatile, proleakproof) a ORDER BY proname, args) TO '/tmp/pg_proc_provolatile_dump.csv' WITH CSV DELIMIER '|' HEADER;
> func TestOverloadsVolatilityMatchesPostgres(t *testing.T) {
> ```

Love this test, very cool.


<!-- Sent from Reviewable.io -->

otan committed:
builtins: categorize the volatilities for each overload

This commit adds volatilities to each overload we implement. For now,
these should match the ones in Postgres whenever possible. Note this is
currently not used anywhere.

We have also added two tests:
* one test that ensures each builtin has a volatility.
* one test that checks the volatilies against a dump of psql
volatilities. We can add a blacklist to this later if we decide to
differ, but for now it is safer to match.

Release note: None

otan commented:
tftr!

bors r=RaduBerinde

jordanlewis commented:
Nice 😭 😭
